### PR TITLE
[New Entry] C: data types 

### LINKED ISSUE #886: [New Entry] C: Data Types
### Reason/inspiration (optional)

Create an entry on Data Types for [C](https://www.codecademy.com/resources/docs/c) language.

_Recommended to search throughout the C entries to see what types are mentioned and link out to this created data types entry. Questions are welcome!_

Let us know if you'd like to take this issue on! Tag someone on the team and ask to be assigned 😸 

### Entry Type (select all that apply)

New Topic Entry

### Code of Conduct

- [X] I agree to follow Codecademy Doc's Code of Conduct

### For Maintainers

- [X] Labels added
- [x] Issue is assigned

### PR DESCRIPTION
**Pull request by Nessvah**
### Description

Closes  #886 

### Type of Change

- Adding a new entry

### Checklist

- [X] All writings are my own.
- [X] My entry follows the Codecademy Docs style guide.
- [X] My changes generate no new warnings.
- [X] I have performed a self-review of my own writing and code.
- [X] I have checked my entry and corrected any misspellings.
- [X] I have made corresponding changes to the documentation if needed.
- [X] I have confirmed my changes are not being pushed from my forked `main` branch.
- [X] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [X] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### CODE REVIEW

File: content/c/concepts/data-types/data-types.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
+---
+Title: 'Data Types'
+Description: 'Java’s most basic data types are known as primitive data types and are in the system by default. The available types are as follows: - int - char - boolean - byte - long'
```
Comment (SSwiniarski) [inline content/c/concepts/data-types/data-types.md:3]:
```suggestion
Description: 'Java’s most basic data types are known as primitive data types and are in the system by default. The available types are as follows: - int - char - boolean - byte - long'
```
```suggestion
Description: 'The most basic data types in C are known as primitive data types and are in the system by default.'
```

Comment (SSwiniarski) [reply content/c/concepts/data-types/data-types.md:3]: Removed reference to Java. Also, descriptions are best as one sentence.

File: content/c/concepts/data-types/data-types.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+```c
+int age = 25;
+float = 125.8;
+double = 3.2325467;
+char = 'T';
+```
+
+### Int
+
+It contains only integral values. A minus sign preceding the data type indicates that the value is negative.
```
Comment (SSwiniarski) [inline content/c/concepts/data-types/data-types.md:39]:
```suggestion
Ints contain only integral values. A minus sign preceding the data type indicates that the value is negative.
```

File: content/c/concepts/data-types/data-types.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+Can be expressed in hexadecimal notation (base 16).
```
Comment (SSwiniarski) [inline content/c/concepts/data-types/data-types.md:40]:
```suggestion
They can be expressed in hexadecimal notation (base 16).
```

Comment (SSwiniarski) [reply content/c/concepts/data-types/data-types.md:40]: I'd add an explanation of the syntax for hexadecimal values.

File: content/c/concepts/data-types/data-types.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+```c
+int = 0;
+int = -5;
+int = 20;
+int rgbColor = 0XFFEF0D;
+```
+
+Spaces aren't allowed between digits and the values larger than 999 don't need a comma.
```
Comment (SSwiniarski) [inline content/c/concepts/data-types/data-types.md:49]:
```suggestion
Spaces aren't allowed between digits and the values larger than 999 don't use commas.
```

File: content/c/concepts/data-types/data-types.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+```c
+int hundreds = 12,000; // This is wrong
+int hundreds2 = 12000 // Correct!
+```
+
+### Float and Double
+
+Floats are used to store floating-point numbers, values that contain decimal plates. They can be expressed in scientific notation and has 6 decimal digits of precision.
```
Comment (SSwiniarski) [inline content/c/concepts/data-types/data-types.md:58]:
```suggestion
Floats are used to store floating-point numbers, values that contain decimal places. They can be expressed in scientific notation and have 6 decimal digits of precision.
```

Comment (SSwiniarski) [reply content/c/concepts/data-types/data-types.md:58]: I'd add a sentence explaining the syntax of scientific notation.

File: content/c/concepts/data-types/data-types.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+Doubles are the same but they are needed when more precision is required since it has 15 decimal digits of precision.
```
Comment (SSwiniarski) [inline content/c/concepts/data-types/data-types.md:60]:
```suggestion
Doubles are the same but they are used when more precision is required. They have 15 decimal digits of precision.
```

File: content/c/concepts/data-types/data-types.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+```c
+float salary = 523.45;
+double bankTransaction = 245.2518;
+float counterFactor = 15.87e4;
+```
+
```
Comment (SSwiniarski) [inline content/c/concepts/data-types/data-types.md:67]: We're missing a section on the `char` datatype.

File: content/c/concepts/data-types/data-types.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+## Memory Size
+
+| Type   | Size    | Value Range                     | Format Specifier |
+| ------ | ------- | ------------------------------- | ---------------- |
+| Char   | 1 byte  | -128 to 127 or 0 to 255         | %c               |
+| Int    | 4 bytes | -2,147,483,648 to 2,147,483,647 | %d               |
+| Float  | 4 bytes | 1.2E-38 to 3.4E+38              | %f               |
+| Double | 8 bytes | 1.7E-308 to 1.7E+308            | %lf              |
```
Comment (SSwiniarski) [inline content/c/concepts/data-types/data-types.md:75]: There's no mention of what the "Format Specifier" means in this entry. That column should probably be omitted.

**Review by SSwiniarski — CHANGES_REQUESTED** [review]
@Nessvah I finished a first review and left some comments.

**Comment by Nessvah** [comment]
> @Nessvah I finished the first review and left some comments.

thank you. I'll make the changes as soon as possible. 
I was considering including how to make the conversions and talk about the signed and unsigned types, what do you think?

**Comment by SSwiniarski** [comment]
> > @Nessvah I finished the first review and left some comments.
> 
> thank you. I'll make the changes as soon as possible. I was considering including how to make the conversions and talk about the signed and unsigned types, what do you think?

@Nessvah, both would be good additions if they're relatively brief. We don't want Docs entries to become overlong.

**Comment by Nessvah** [comment]
> > > @Nessvah I finished the first review and left some comments.
> > 
> > 
> > thank you. I'll make the changes as soon as possible. I was considering including how to make the conversions and talk about the signed and unsigned types, what do you think?
> 
> @Nessvah, both would be good additions if they're relatively brief. We don't want Docs entries to become overlong.

@SSwiniarski  Let me know if after this review there's still space to improve and maybe add those two topics.

File: content/c/concepts/data-types/data-types.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
+---
+Title: 'Data Types'
+Description: 'C’s most basic data types are known as primitive data types and are in the system by default.'
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:3]:
```suggestion
Description: 'Data types are units of value stored in memory through variables, each with different functionalities and size-ranges.'
```

File: content/c/concepts/data-types/data-types.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+int age = 25;
+float = 125.8;
+double = 3.2325467;
+char = 'T';
+```
+
+### Int
+
+Ints contains only integral values. A minus sign preceding the data type indicates that the value is negative.
+They can be expressed in hexadecimal notation (base 16).\*
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:40]:
```suggestion
Ints contains only integral values. A minus sign preceding the data type indicates that the value is negative. They can be expressed in hexadecimal notation (base 16).
```

File: content/c/concepts/data-types/data-types.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+\* Hexadecimal integer literals begin always with the 0 followed by either an `x` or `X`, followed by different combinations of digits going from `0` through `9`, and the letters `a` through `f` or `A` through `F`.
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:42]:
```suggestion
> **Note:** Hexadecimal integer literals begin always with the `0` followed by either an `x` or `X`, followed by different combinations of digits (`0` - `9`) and letters (`a` - `f` or `A` - `F`).
```

File: content/c/concepts/data-types/data-types.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+int = -5;
+int = 20;
+int rgbColor = 0XFFEF0D; // Hexadecimal
+```
+
+Spaces aren't allowed between digits and the values larger than 999 don't use commas.
+
+```c
+int hundreds = 12,000; // This is wrong
+int hundreds2 = 12000 // Correct!
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:55]:
```suggestion
int hundreds2 = 12000; // This is correct
```

File: content/c/concepts/data-types/data-types.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+```
+
+### Float and Double
+
+Floats are used to store floating-point numbers, values that contain decimal places. They can be expressed in scientific notation\* and has 6 decimal digits of precision.
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:60]:
```suggestion
Floating-point values are non-whole numbers that contain decimal places. They can have up to 6 decimal digits of precision and can be expressed in scientific notation.
```

File: content/c/concepts/data-types/data-types.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+\*Scientific notation is a form of expressing numbers that are too large or too small to be conveniently written in decimal form. E.g. `6.022E23`
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:62]:
```suggestion
> **Note:** Scientific notation is a form of expressing numbers that are too large or too small to be conveniently written in decimal form ( e.g., `6.022E23`).
```

File: content/c/concepts/data-types/data-types.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+Doubles are the same but they are used when more precision is required. They have 15 decimal digits of precision.
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:64]:
```suggestion
Doubles are like floats but more precise with up to 15 decimal digits. The following is an example of `float` and `double` type values:
```

File: content/c/concepts/data-types/data-types.md [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+```c
+float salary = 523.45;
+double bankTransaction = 245.2518;
+float counterFactor = 15.87e4;
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:69]:
```suggestion
float counterFactor = 15.87E4;
```

Comment (Dusch4593) [reply content/c/concepts/data-types/data-types.md:69]: Not sure if capitalization matters but suggesting uppercase "E" for consistency.

File: content/c/concepts/data-types/data-types.md [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+## Memory Size
+
+| Type   | Size    | Value Range                     |
+| ------ | ------- | ------------------------------- |
+| Char   | 1 byte  | -128 to 127 or 0 to 255         |
+| Int    | 4 bytes | -2,147,483,648 to 2,147,483,647 |
+| Float  | 4 bytes | 1.2E-38 to 3.4E+38              |
+| Double | 8 bytes | 1.7E-308 to 1.7E+308            |
+
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:89]: I would remove this section altogether and, instead, place this table right above the "### Int" header.

File: content/c/concepts/data-types/data-types.md [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+```c
+float salary = 523.45;
+double bankTransaction = 245.2518;
+float counterFactor = 15.87e4;
+```
+
+### Char
+
+Chars allows its variables to store only a single character a letter, number, space, punctuation mark or symbol. Single quotes are used to represent them. They only occupy one byte in memory.
+In C, a sequence of chars are called [strings](https://www.codecademy.com/resources/docs/c/strings).
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:75]:
```suggestion
Chars use single quotation marks `'` to present single characters such as letters, [numbers](https://www.codecademy.com/resources/docs/general/number), [whitespace](https://www.codecademy.com/resources/docs/general/whitespace), special characters, and symbols. Each char only occupies one byte in memory. In C, a sequence of chars are called [strings](https://www.codecademy.com/resources/docs/c/strings).
```

File: content/c/concepts/data-types/data-types.md [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+| Type   | Size    | Value Range                     |
+| ------ | ------- | ------------------------------- |
+| Char   | 1 byte  | -128 to 127 or 0 to 255         |
+| Int    | 4 bytes | -2,147,483,648 to 2,147,483,647 |
+| Float  | 4 bytes | 1.2E-38 to 3.4E+38              |
+| Double | 8 bytes | 1.7E-308 to 1.7E+308            |
+
+## Static Typing
+
+C is a statically-typed language (such as C++ and Java) where variable types are known at compile time instead of the runtime. In this case, types must be expressly indicated by the programmer.
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:92]: This section should be moved somewhere near the top of this file (in lines # 15-18).

**Review by Dusch4593 — CHANGES_REQUESTED** [review]
@Nessvah Finished with my review and left some comments/suggestions. The content in this entry seems to only cover _primitive data types_. But we should also include information about these other classifications:

- Derived types ([arrays](https://www.codecademy.com/resources/docs/c/arrays), [functions](https://www.codecademy.com/resources/docs/c/functions), [pointers](https://www.codecademy.com/resources/docs/c/pointers))
- User defined types ([structures](https://www.codecademy.com/resources/docs/c/structures))

File: content/c/concepts/data-types/data-types.md [hunk 20]
```diff
@@ -0,0 +1,92 @@
+---
+Title: 'Data Types'
+Description: 'C’s most basic data types are known as primitive data types and are in the system by default.'
+Subjects:
+  - 'Computer Science'
+Tags:
+  - 'Data Types'
+  - 'Integers'
+  - 'Booleans'
```
[the remark this answers is not in the document]
Comment (Nessvah) [reply content/c/concepts/data-types/data-types.md:9]: I should always write the tags in order, right? For a matter of consistency?

[the remark this answers is not in the document]
Comment (Dusch4593) [reply content/c/concepts/data-types/data-types.md:9]: Nope! Order does not matter 😄

File: content/c/concepts/data-types/data-types.md [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+
+- Primitive data types
+- User-defined data types
+- Derived data types
+
+Types also differ in the amount of memory they occupy and the range of values they can hold. This will depend on the compiler of the computer. An integer, for example, might take up 32 bits on one computer, or perhaps it might be stored in 64 bits on another.
+
+## Primitive types
+
+These data types are mostly numeric (integers, doubles, floating-point numbers, etc.) and can be used for arithmetic operations like subtraction and division. Values that use the `void` keyword (for no data) are also considered to be primitive data types.
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:32]:
```suggestion
These data types are mostly numeric (integers, doubles, floating-point numbers, etc.) and can be used for arithmetic operations like subtraction and division. Values that use the `void` keyword (for no data) are also regarded as primitive data types.
```

File: content/c/concepts/data-types/data-types.md [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+
+```c
+float salary = 523.45;
+double bankTransaction = 245.2518;
+float counterFactor = 15.87E4;
+```
+
+### Char
+
+Chars use single quotations marks `'` to present single characters such as letters, [numbers](https://www.codecademy.com/resources/docs/general/number), [whitespace](https://www.codecademy.com/resources/docs/general/whitespace), special characters, and symbols. Each char only occupies one byte in memory. In C, a sequence of chars is called [strings](https://www.codecademy.com/resources/docs/c/strings).
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:96]:
```suggestion
Chars use single quotations marks `'` to represent single characters such as letters, [numbers](https://www.codecademy.com/resources/docs/general/number), [whitespace](https://www.codecademy.com/resources/docs/general/whitespace), special characters, and symbols. Each char only occupies one byte in memory. In C, a sequence of chars is called [strings](https://www.codecademy.com/resources/docs/c/strings).
```

File: content/c/concepts/data-types/data-types.md [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+Tags:
+  - 'Booleans'
+  - 'Variable Types'
+  - 'Variables'
+CatalogContent:
+  - 'learn-c'
+  - 'paths/computer-science'
+---
+
+In C, **data types** are units of value stored in memory through [variables](https://www.codecademy.com/resources/docs/c/variables). Different kinds of data types determine their size and functionality.
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:16]:
```suggestion
In C, **data types** are units of value stored in [memory](https://www.codecademy.com/resources/docs/c/memory-management) through [variables](https://www.codecademy.com/resources/docs/c/variables). Different kinds of data types determine their size and functionality.
```

File: content/c/concepts/data-types/data-types.md [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+- Union
+
+## Memory Size
+
+| Type   | Size    | Value Range                     |
+| ------ | ------- | ------------------------------- |
+| Char   | 1 byte  | -128 to 127 or 0 to 255         |
+| Int    | 4 bytes | -2,147,483,648 to 2,147,483,647 |
+| Float  | 4 bytes | 1.2E-38 to 3.4E+38              |
+| Double | 8 bytes | 1.7E-308 to 1.7E+308            |
```
Comment (Dusch4593) [inline content/c/concepts/data-types/data-types.md:58]:
```suggestion
| Char   | 1 byte  | -128 - 127 or 0 - 255           |
| Int    | 4 bytes | -2,147,483,648 - 2,147,483,647  |
| Float  | 4 bytes | 1.2E-38 - 3.4E+38               |
| Double | 8 bytes | 1.7E-308 - 1.7E+308             |
```

**Review by Dusch4593 — CHANGES_REQUESTED** [review]
@Nessvah Just a few more comments/suggestions from me 😄

**Comment by Nessvah** [comment]
> @Nessvah Just a few more comments/suggestions from me smile

😃 Awesome! I'll take a look at it later today.

File: content/c/concepts/data-types/data-types.md [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+CatalogContent:
+  - 'learn-c'
+  - 'paths/computer-science'
+---
+
+In C, **data types** are units of value stored in [memory](https://www.codecademy.com/resources/docs/c/memory-management) through [variables](https://www.codecademy.com/resources/docs/c/variables). Different kinds of data types determine their size and functionality.
+
+## Static Typing
+
+Like with C++ and Java, C is a statically-typed language where variable types are known at compile time instead of the [runtime](https://www.codecademy.com/resources/docs/general/runtime). In this case, types must be expressly indicated by the programmer.
```
Comment (yangc95) [inline content/c/concepts/data-types/data-types.md:20]: Do we mean that types are explicitly indicated?

File: content/c/concepts/data-types/data-types.md [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+
+```c
+float salary = 523.45;
+double bankTransaction = 245.2518;
+float counterFactor = 15.87E4;
+```
+
+### Char
+
+Chars use single quotations marks `'` to represent single characters such as letters, [numbers](https://www.codecademy.com/resources/docs/general/number), [whitespace](https://www.codecademy.com/resources/docs/general/whitespace), special characters, and symbols. Each char only occupies one byte in memory. In C, a sequence of chars is called [strings](https://www.codecademy.com/resources/docs/c/strings).
```
Comment (SSwiniarski) [inline content/c/concepts/data-types/data-types.md:96]:
```suggestion
Chars use single quotations marks `'` to represent single characters such as letters, [numbers](https://www.codecademy.com/resources/docs/general/number), [whitespace](https://www.codecademy.com/resources/docs/general/whitespace), special characters, and symbols. Each char only occupies one byte in memory. In C, sequences of chars are called [strings](https://www.codecademy.com/resources/docs/c/strings).
```

**Review by SSwiniarski — APPROVED** [review]
@Nessvah Looks good to me, I just had one additional comment.

**Review by Dusch4593 — APPROVED** [review]
@Nessvah LGTM! 🚀

**Comment by Dusch4593** [comment]
Congrats and thank you so much for another awesome contribution to Codecademy Docs, @Nessvah! 🚀

The entry should now be live: https://www.codecademy.com/resources/docs/c/data-types

Screen Shot 2022-10-31 at 2 34 41 PM

Looking forward to your next contribution! 😄

**Comment by Nessvah** [comment]
@Dusch4593 Thank you so much! It's always a pleasure to help and learn with you guys 😄 👍🏻

**Pull request merged**